<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Ryan Nickens – rynickens@gmail.com" />
<meta name="author" content="Livia Tran – livia.tran@ucsf.edu" />
<meta name="author" content="Leamon Crooms IV – lcroomsiv@gmail.com" />
<meta name="author" content="Derek Pappas, PhD – djpappas75@gmail.com" />
<meta name="author" content="Vinh Luu – vinhluu864@berkeley.edu" />
<meta name="author" content="Josh Bredeweg" />
<meta name="author" content="Steven J. Mack, PhD – steven.mack@ucsf.edu" />

<meta name="date" content="2024-07-21" />

<title>HLAtools</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">HLAtools</h1>
<h4 class="author">Ryan Nickens – <a href="mailto:rynickens@gmail.com" class="email">rynickens@gmail.com</a></h4>
<h4 class="author">Livia Tran – <a href="mailto:livia.tran@ucsf.edu" class="email">livia.tran@ucsf.edu</a></h4>
<h4 class="author">Leamon Crooms IV – <a href="mailto:lcroomsiv@gmail.com" class="email">lcroomsiv@gmail.com</a></h4>
<h4 class="author">Derek Pappas, PhD – <a href="mailto:djpappas75@gmail.com" class="email">djpappas75@gmail.com</a></h4>
<h4 class="author">Vinh Luu – <a href="mailto:vinhluu864@berkeley.edu" class="email">vinhluu864@berkeley.edu</a></h4>
<h4 class="author">Josh Bredeweg</h4>
<h4 class="author">Steven J. Mack, PhD – <a href="mailto:steven.mack@ucsf.edu" class="email">steven.mack@ucsf.edu</a></h4>
<h4 class="date">2024-07-21</h4>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(HLAtools)</span></code></pre></div>
<ul>
<li>Package Version: 1.1.1</li>
</ul>
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>The <em>HLAtools</em> package contains a set of computable resources
and a suite of tools intended to facilitate the effective application
and analysis of the named genes in the HLA region.</p>
<p>HLAtools includes:</p>
<ul>
<li>Reference datasets that describe the genes supported by the
IPD-IMGT/HLA Database, organize them into useful categories, and
identify their gene feature boundaries.</li>
<li>The ability to build local amino acid, codon, coding nucleotide, and
genomic nucleotide sequence alignments for HLA genes of interest.</li>
<li>Advanced tools of the comparison and dissection of allele sequences,
and the construction of customized multi-gene sequence alignments.</li>
<li>Functions that translate allele names, vectors and data frames of
allele names, and Genotype List (GL) String Code messages across
IPD-IMGT/HLA database release versions, and inter-convert between GL
String and UNIFORMAT formats.</li>
<li>Data-analysis tools that extend the utility of Bridging
Immunogenomic Data-Analysis Workflow Gaps (BIGDAWG) formatted datasets
by performing mutliallele and multilocus stratification, calculating
Relative Risk measures, and converting BIGDAWG datasets to PyPop
datasets.</li>
</ul>
<hr />
</div>
<div id="reference-datasets" class="section level1">
<h1>1. Reference Datasets</h1>
<div id="atlas-of-gene-feature-boundaries" class="section level2">
<h2>Atlas of Gene Feature Boundaries</h2>
<p>The <em>HLAatlas</em> data object is a list object of sub-lists of R
data frames identifying the location of boundary positions between
gene-features (exons, introns and untranslated regions [UTRs])
identified in the protein, nucleotide and genomic alignments for each
gene supported in the <a href="https://github.com/ANHIG/IMGTHLA/tree/Latest/alignments">ANHIG/IMGTHLA
GitHub Repository</a>. Example protein (<em>prot</em>), nucleotide
(<em>nuc</em>) and genomic (<em>gen</em>) atlases for HLA-A are shown
below. A new <em>HLAatlas</em> can be built after each IPD-IMGT/HLA
Database release using the <em>updateAll()</em> function, although the
atlases are not expected to change unless new genes are added to the
IPD-IMGT/HLA Database.</p>
<p>Building the <em>HLAatlas</em> object requires internet access, and
it can take several minutes to build a full complement of atlases. For
example, building all of the atlases for IPD-IMGT/HLA Database release
version 3.56.0 took 4.2 minutes on a 3.3 GHz 12-Core Intel Xenon W 2019
Mac Pro with 288 GB of 2933 MHZ DDR4 RAM, and 4.6 minutes on a 2.4 GHz
8-Core Intel Core i9 2019 MacBook Pro with 64 GB of 2667 MHZ DDR4 RAM,
and 5.24 minutes on a 2.4 GHz Intel Core i9-10885H Dell Precision 5500
with 32 GB of RAM.</p>
<p><em>HLAatlas</em> objects can be built for all HLA region genes with
available genomic, nucleotide and amino acid alignments in release
versions 3.00.0 to 3.57.0. Barring changes in the structure of the
source data for this object in future IPD-IMGT/HLA Database releases, it
should be possible to build <em>HLAatlas</em> objects for future
releases.</p>
<div id="protein-sequence-atlases" class="section level3">
<h3>Protein Sequence Atlases</h3>
<p>The column headers for <em>prot</em> atlases identify the peptide
residues encoded by codons that follow or contain an exon (E) boundary.
The HLA-A <em>prot</em> atlas below illustrates that the codon encoding
amino acid 1 included the Exon 1: Exon 2 boundary (E.1-2).</p>
<pre><code>HLAatlas$prot$A
   E.1-2 E.2-3 E.3-4 E.4-5 E.5-6 E.6-7 E.7-8
AA     1    91   183   275   314   325   341</code></pre>
</div>
<div id="nucleotide-sequence-atlases" class="section level3">
<h3>Nucleotide Sequence Atlases</h3>
<p>The column headers for <em>nuc</em> atlases identify the cDNA and
codon positions preceded by an exon (E) boundary. The HLA-A <em>nuc</em>
atlas below illustrates that the Exon 1:Exon 2 boundary (E.1-2) is
between nucleotide positions 73 and 74 and within codon 1.</p>
<pre><code>HLAatlas$nuc$A
      E.1-2 E.2-3 E.3-4 E.4-5 E.5-6 E.6-7 E.7-8
cDNA     74   344   620   896  1013  1046  1094
codon     1    91   183   275   314   325   341</code></pre>
</div>
<div id="genomic-sequence-atlases" class="section level3">
<h3>Genomic Sequence Atlases</h3>
<p>The column headers for <em>gen</em> atlases identify the genomic
nucleotide positions preceded by a UTR (U), exon (E) or intron (I)
boundary. The HLA-A <em>gen</em> atlas below illustrates that the
boundary between the 5’ UTR (U.5) and Exon 1 (E.1) is between genomic
positions -1 and 1, and that the boundary between Intron 3 and Exon 4 is
between positions 1569 and 1570 (presented as a table for
readability).</p>
<table style="width:100%;">
<colgroup>
<col width="3%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"></th>
<th align="center">U.5-E.1</th>
<th align="center">E.1-I.1</th>
<th align="center">I.1-E.2</th>
<th align="center">E.2-I.2</th>
<th align="center">I.2-E.3</th>
<th align="center">E.3-I.3</th>
<th align="center">I.3-E.4</th>
<th align="center">E.4-I.4</th>
<th align="center">I.4-E.5</th>
<th align="center">E.5-I.5</th>
<th align="center">I.5-E.6</th>
<th align="center">E.6-I.6</th>
<th align="center">I.6-E.7</th>
<th align="center">E.7-I.7</th>
<th align="center">I.7-E.8</th>
<th align="center">E.8-U.3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">gDNA</td>
<td align="center">1</td>
<td align="center">74</td>
<td align="center">204</td>
<td align="center">474</td>
<td align="center">715</td>
<td align="center">991</td>
<td align="center">1570</td>
<td align="center">1846</td>
<td align="center">1948</td>
<td align="center">2065</td>
<td align="center">2507</td>
<td align="center">2540</td>
<td align="center">2682</td>
<td align="center">2730</td>
<td align="center">2899</td>
<td align="center">2904</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="historical-catalogue-of-hla-allele-names" class="section level2">
<h2>Historical Catalogue of HLA Allele Names</h2>
<p>The <em>alleleListHistory</em> data object is a data frame that
identifies all HLA and MIC allele names and their accession identifiers
(HLA_IDs) for all IPD-IMGT/HLA Database release versions (e.g., “X3350”
represents version 3.35.0) going back to version 1.05.0 (January of
2000), and includes allele names that have been changed or deleted. A
new <em>alleleListHistory</em> can be built after each IPD-IMGT/HLA
Database release using the <em>updateAll()</em> function.</p>
<pre><code>alleleListHistory$AlleleListHistory[1:5,1:5]
    HLA_ID          X3560          X3550          X3540          X3530
1 HLA00001  A*01:01:01:01  A*01:01:01:01  A*01:01:01:01  A*01:01:01:01
2 HLA00002  A*01:02:01:01  A*01:02:01:01  A*01:02:01:01  A*01:02:01:01
3 HLA00003  A*01:03:01:01  A*01:03:01:01  A*01:03:01:01  A*01:03:01:01
4 HLA00004 A*01:04:01:01N A*01:04:01:01N A*01:04:01:01N A*01:04:01:01N
5 HLA00005  A*02:01:01:01  A*02:01:01:01  A*02:01:01:01  A*02:01:01:01

alleleListHistory$AlleleListHistory[c(65,100:105,2094),c(1,23,58:59,81,96,101,102)]
       HLA_ID         X3350      X3000     X2280  X2090  X1110  X1060  X1050
65   HLA00069       A*24:19    A*24:19    A*2419 A*2419 A*2419   &lt;NA&gt;   &lt;NA&gt;
100  HLA00107       A*33:04    A*33:04    A*3304 A*3304 A*3304 A*3304 A*3304
101  HLA00108 A*34:01:01:01 A*34:01:01  A*340101 A*3401 A*3401 A*3401 A*3401
102  HLA00109 A*34:02:01:01    A*34:02    A*3402 A*3402 A*3402 A*3402 A*3402
103  HLA00110       A*36:01    A*36:01    A*3601 A*3601 A*3601 A*3601 A*3601
104  HLA00111       A*43:01    A*43:01    A*4301 A*4301 A*4301 A*4301 A*4301
105  HLA00112 A*66:01:01:01    A*66:01    A*6601 A*6601 A*6601 A*6601 A*6601
2094 HLA02186     MICB*021N  MICB*021N MICB*021N   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;</code></pre>
</div>
<div id="functional-and-organizational-categories-of-genes-in-the-hla-region" class="section level2">
<h2>Functional and Organizational Categories of Genes in the HLA
Region</h2>
<div id="molecular-characteristics-of-hla-region-genes" class="section level3">
<h3>Molecular Characteristics of HLA Region Genes</h3>
<p>The <em>IMGTHLAGeneTypes</em> data object describes the <a href="https://hla.alleles.org/genes/index.html">named genes in the HLA
region</a> curated by the IPD-IMGT/HLA Database.
<em>IMGTHLAGeneTypes</em> distinguishes pseudogenes and gene fragments
from expressed genes, and summarizes each gene’s molecular
characteristics.</p>
<p>The information in this object can be found online at <a href="https://hla.alleles.org/genes/index.html">https://hla.alleles.org/genes/index.html</a>.
A new <em>IMGTHLAGeneTypes</em> can be built using the
<em>updateAll()</em> function, although the source data is not expected
to change unless new genes are added to the IPD-IMGT/HLA Database.</p>
</div>
<div id="gazeeteer-of-hla-region-genes" class="section level3">
<h3>Gazeeteer of HLA Region Genes</h3>
<p>The <em>HLAgazeteer</em> data object is a list object that organizes
the HLA region genes supported by the IPD-IMGT/HLA Database in nineteen
vectors describing the availability of alignments, gene functionality,
group identity and map order. A new <em>HLAgazeteer</em> can be built
after each IPD-IMGT/HLA Database release using the <em>updateAll()</em>
function, although the <em>HLAgazeteer</em> is not expected to change
unless new genes are added to the IPD-IMGT/HLA Database.</p>
<pre><code>names(HLAgazeteer)
 [1] &quot;align&quot; &quot;gen&quot; &quot;nuc&quot; &quot;prot&quot; &quot;nogen&quot; &quot;nonuc&quot; &quot;noprot&quot; &quot;pseudo&quot; &quot;frag&quot; &quot;hla&quot; &quot;expressed&quot; &quot;notexpressed&quot; &quot;classireg&quot; &quot;classihla&quot; &quot;classiireg&quot; &quot;classiihla&quot; &quot;classical&quot; &quot;nonclassical&quot; &quot;map&quot; &quot;version&quot;     </code></pre>
<p>For example, the <em>$align</em> vector includes all of the genes for
which <strong>sequence alignments</strong> are available:</p>
<pre><code>HLAgazeteer$align
 [1] &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;DMA&quot; &quot;DMB&quot; &quot;DOA&quot; &quot;DOB&quot; &quot;DPA1&quot; &quot;DPA2&quot; &quot;DPB1&quot; &quot;DPB2&quot; &quot;DQA1&quot; &quot;DQA2&quot; &quot;DQB1&quot; &quot;DQB2&quot; &quot;DRA&quot;  &quot;DRB1&quot; &quot;DRB2&quot; &quot;DRB3&quot; &quot;DRB4&quot; &quot;DRB5&quot; &quot;DRB6&quot; &quot;DRB7&quot; &quot;DRB8&quot; &quot;DRB9&quot; &quot;E&quot; &quot;F&quot; &quot;G&quot; &quot;H&quot; &quot;HFE&quot; &quot;J&quot; &quot;K&quot; &quot;L&quot; &quot;MICA&quot; &quot;MICB&quot; &quot;N&quot; &quot;P&quot; &quot;S&quot; &quot;T&quot; &quot;TAP1&quot; &quot;TAP2&quot; &quot;U&quot; &quot;V&quot; &quot;W&quot; &quot;Y&quot;     </code></pre>
<p>The <em>$prot</em> vector includes all of the genes with
<strong>protein</strong> alignments:</p>
<pre><code>HLAgazeteer$prot
 [1] &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;DMA&quot; &quot;DMB&quot; &quot;DOA&quot; &quot;DOB&quot; &quot;DPA1&quot; &quot;DPB1&quot; &quot;DQA1&quot; &quot;DQA2&quot; &quot;DQB1&quot; &quot;DQB2&quot; &quot;DRA&quot; &quot;DRB1&quot; &quot;DRB3&quot; &quot;DRB4&quot; &quot;DRB5&quot; &quot;DRB&quot; &quot;E&quot; &quot;F&quot; &quot;G&quot; &quot;HFE&quot; &quot;MICA&quot; &quot;MICB&quot; &quot;TAP1&quot; &quot;TAP2&quot;</code></pre>
<p>Note that the <em>$prot</em> and <em>$nuc</em> vectors include a
‘DRB’ “gene”. While ‘DRB’ is not a gene name, the DRB_prot.txt file
includes combined alignments for the DRB1, DRB3, DRB4, and DRB5 genes,
and the DRB_nuc.txt file includes combined alignments for the DRB1,
DRB2, DRB3, DRB4, DRB5, DRB6, DRB7, DRB8, and DRB9 genes. ‘DRB’ is
included in these vectors for the purpose of validation when these
combined alignments are desired.</p>
</div>
</div>
<div id="annotation-of-pseudogene-and-gene-fragement-features" class="section level2">
<h2>Annotation of Pseudogene and Gene Fragement Features</h2>
<p>The <em>fragmentFeatureNames</em> object identifies and annotates the
non-standard gene features found in pseudogenes and gene fragments,
based on the positions of feature boundaries (“|”) in the pertinent
genomic alignment sequence. Where the features of functional genes are
limited to Introns (I), Exons (E) and Untranslated Regions (U), the
non-standard features in pseudogenes and gene fragments are described
as:</p>
<ul>
<li>Joins (J), when a feature includes sequence from two or more
sequences that are separated by a boundary in the reference</li>
<li>Segments (S), when the sequence of a feature is a subset of a longer
feature sequence</li>
<li>Novel (N), when the sequence of a feature does not correspond to a
known sequence</li>
<li>Hybrid (H), when a feature includes a sequence to belongs to a known
feature as well as nucleotide sequence that does not correspond to a
feature</li>
</ul>
<p>For each pseudogene or gene fragment, the
<em>fragmentFeatureNames</em> element contains a $features element
identifying the gene features in the 5’ to 3’ direction, and an
$annotation element that provides some detail about the feature.</p>
<p>These non-standard gene features are included the feature boundaries
of the gene fragments and pseudogenes in the <em>HLAatlas</em>.</p>
<p>The current annotations should not change across IPD-IMGT/HLA
Database releases. Wen a new pseudogene or gene-fragment is added to the
IPD-IMGT/HLA Database, a new annotation will be added as part of a
package update.</p>
<pre><code>fragmentFeatureNames$DPA2
$features
[1] &quot;U.5&quot; &quot;E.1&quot; &quot;I.1&quot; &quot;E.2&quot; &quot;I.2&quot; &quot;E.3&quot; &quot;I.3&quot; &quot;E.4&quot; &quot;U.3&quot;
$annotation
[1] &quot;All of the reference gene features are present. E.1 starts 28 nucleotides before the reference, and ends 88 nucleotides before the reference.&quot;

fragmentFeatureNames$L
$features
 [1] &quot;U.5&quot; &quot;E.1&quot; &quot;I.1&quot; &quot;E.2&quot; &quot;I.2&quot; &quot;E.3&quot; &quot;I.3&quot; &quot;E.4&quot; &quot;I.4&quot; &quot;E.5&quot; &quot;I.5&quot; &quot;E.6&quot; &quot;I.6&quot; &quot;E.7&quot; &quot;I.7&quot; &quot;E.8&quot; &quot;U.3&quot;
$annotation
[1] &quot;All of the reference gene features are present.&quot;

fragmentFeatureNames$P
$features
 [1] &quot;J.1&quot; &quot;E.3&quot; &quot;I.3&quot; &quot;E.4&quot; &quot;I.4&quot; &quot;E.5&quot; &quot;I.5&quot; &quot;E.6&quot; &quot;I.6&quot; &quot;E.7&quot; &quot;I.7&quot; &quot;U.3&quot;
$annotation
[1] &quot;J.1 is ~350 nucleotides of novel sequence followed by ~120 nucleotides from the 5&#39; end of Intron 2. Exon 8 is absent.&quot;

fragmentFeatureNames$S
$features
[1] &quot;H.1&quot; &quot;S.1&quot; &quot;J.1&quot; &quot;E.7&quot; &quot;I.7&quot; &quot;E.9&quot; &quot;S.2&quot;
$annotation
[1] &quot;H.1 is 37 nucleotides of novel sequence, followed by the last 185 nucleotides of Intron 5. S.1 is the first 27 nucleotides of Exon 6. J.1 is the last 4 nucleotides of Exon 6 (2 nucleotides in the reference have been deleted), followed by the last 100 nucleotides of Intron 6 (6 nucleotides in the reference have been deleted). E.9 is a 191 nucleotide-long Exon in what is the 5&#39; end of the 3&#39; UTR in the reference. S.2 is the 3&#39; end of the 3&#39; UTR.&quot;</code></pre>
</div>
<div id="sequence-alignments" class="section level2">
<h2>Sequence Alignments</h2>
<p>The <em>HLAalignments</em> data object is a list object of sub-lists
of data frames of peptide (<em>prot</em>), codon (<em>codon</em>),
coding nucleotide (<em>nuc</em>) and genomic nucleotide (<em>gen</em>)
alignments for the HLA and HLA-region genes supported in the <a href="https://github.com/ANHIG/IMGTHLA/tree/Latest/alignments">ANHIG/IMGTHLA
GitHub Repository</a>, along with a <em>version</em> character string
that identifies the IPD-IMGT/HLA Database release under which the
<em>HLAalignments</em> object was built.</p>
<p>Given the size of all the combined alignments, <em>HLAalignments</em>
<strong>is not bundled with the HLAtools package</strong>. All
alignments, or desired subsets of each type of alignment for a given
gene, can be built using the <em>alignmentFull()</em> function, as
described in Section 2 <em>Working with Sequence Alignments</em>, below.
<em>HLAalignments</em> objects can be built for all HLA region genes
with available genomic, nucleotide and amino acid alignments in
IPD-IMGT/HLA release versions 3.00.0 to 3.57.0. Barring changes in the
structure of the source data for this object in future IPD-IMGT/HLA
Database releases, it should be possible to build <em>HLAalignments</em>
objects for future releases.</p>
<p>These protein, codon, nucleotide and genomic DNA sequence alignments
are described in detail below.</p>
<p>The first four columns of each alignment identify the locus, allele,
two-field allele name (trimmed_allele), and full allele name
(allele_name) for each sequence, while the remaining columns identify
the sequence at each position, as illustrated below.</p>
<div id="protein-aa-aligments" class="section level3">
<h3>Protein (AA) Aligments</h3>
<p>The column header for the <em>prot</em> alignments identifies
individual peptide positions, starting from the first peptide of the
leader sequence (a negative position) or the first peptide of the native
protein when there is no leader sequence. For example, as shown below,
the HLA-A <em>prot</em> alignment begins at residue -24.</p>
<pre><code>HLAalignments$prot$A[1:5,1:15]
  locus       allele trimmed_allele    allele_name -24 -23 -22 -21 -20 -19 -18 -17 -16 -15 -14
1     A  01:01:01:01        A*01:01  A*01:01:01:01   M   A   V   M   A   P   R   T   L   L   L
2     A 01:01:01:02N        A*01:01 A*01:01:01:02N   M   A   V   M   A   P   R   T   L   L   L
3     A  01:01:01:03        A*01:01  A*01:01:01:03   M   A   V   M   A   P   R   T   L   L   L
4     A  01:01:01:04        A*01:01  A*01:01:01:04   M   A   V   M   A   P   R   T   L   L   L
5     A  01:01:01:05        A*01:01  A*01:01:01:05   M   A   V   M   A   P   R   T   L   L   L</code></pre>
<p>In contrast, the TAP <em>prot</em> alignment begins at residue 1, as
shown below.</p>
<pre><code>HLAalignments$prot$TAP1[1:5,1:15]
  locus      allele trimmed_allele      allele_name 1 2 3 4 5 6 7 8 9 10 11
1  TAP1 01:01:01:01     TAP1*01:01 TAP1*01:01:01:01 M A S S R C P A P  R  G
2  TAP1 01:01:01:02     TAP1*01:01 TAP1*01:01:01:02 M A S S R C P A P  R  G
3  TAP1 01:01:01:03     TAP1*01:01 TAP1*01:01:01:03 M A S S R C P A P  R  G
4  TAP1 01:01:01:04     TAP1*01:01 TAP1*01:01:01:04 M A S S R C P A P  R  G
5  TAP1 01:01:01:05     TAP1*01:01 TAP1*01:01:01:05 M A S S R C P A P  R  G</code></pre>
</div>
<div id="codon-and-individual-nucleotide-cdna-sequence-alignments" class="section level3">
<h3>Codon and Individual Nucleotide (cDNA) Sequence Alignments</h3>
<p>Codon-triplet (<em>codon</em>) and individual nucleotide
(<em>nuc</em>) alignments are included as separate data frames.</p>
<p>The column heads for the <em>codon</em> alignments identify the
individual nucleotide positions within each codon, starting with the
first codon. As shown below, the HLA-A <em>codon</em> alignment begins
at the first nucleotide in codon position -24, which is identified as
<em>-24</em>; the second nucleotide in codon -24 is identified as
<em>-24.1</em> and the third nucleotide in codon -24 is identified as
<em>-24.2</em>.</p>
<pre><code>HLAalignments$codon$A[1:5,1:12]
  locus       allele trimmed_allele    allele_name -24 -24.1 -24.2 -23 -23.1 -23.2 -22 -22.1
1     A  01:01:01:01        A*01:01  A*01:01:01:01   A     T     G   G     C     C   G     T
2     A 01:01:01:02N        A*01:01 A*01:01:01:02N   A     T     G   G     C     C   G     T
3     A  01:01:01:03        A*01:01  A*01:01:01:03   A     T     G   G     C     C   G     T
4     A  01:01:01:04        A*01:01  A*01:01:01:04   A     T     G   G     C     C   G     T
5     A  01:01:01:05        A*01:01  A*01:01:01:05   A     T     G   G     C     C   G     T</code></pre>
<p>The column heads for the <em>nuc</em> aligmnents identify each
individual nucleotide position, starting from the first transcribed
nucleotide (1), as shown below.</p>
<pre><code>HLAalignments$nuc$A[1:5,1:21]
  locus       allele trimmed_allele    allele_name 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17
1     A  01:01:01:01        A*01:01  A*01:01:01:01 A T G G C C G T C  A  T  G  G  C  G  C  C
2     A 01:01:01:02N        A*01:01 A*01:01:01:02N A T G G C C G T C  A  T  G  G  C  G  C  C
3     A  01:01:01:03        A*01:01  A*01:01:01:03 A T G G C C G T C  A  T  G  G  C  G  C  C
4     A  01:01:01:04        A*01:01  A*01:01:01:04 A T G G C C G T C  A  T  G  G  C  G  C  C
5     A  01:01:01:05        A*01:01  A*01:01:01:05 A T G G C C G T C  A  T  G  G  C  G  C  C</code></pre>
</div>
<div id="genomic-gdna-sequence-alignments" class="section level3">
<h3>Genomic (gDNA) Sequence Alignments</h3>
<p>The column heads for the <em>gen</em> alignments identify the
individual nucleotide positions in the ‘full gene’ sequence, which
usually starts in the 5’ untranslated region (UTR) sequence for
expressed genes.</p>
<pre><code>HLAalignments$gen$A[1:5,1:13]
  locus       allele trimmed_allele    allele_name -300 -299 -298 -297 -296 -295 -294 -293 -292
1     A  01:01:01:01        A*01:01  A*01:01:01:01    C    A    G    G    A    G    C    A    G
2     A 01:01:01:02N        A*01:01 A*01:01:01:02N    *    *    *    *    *    *    *    *    *
3     A  01:01:01:03        A*01:01  A*01:01:01:03    C    A    G    G    A    G    C    A    G
4     A  01:01:01:04        A*01:01  A*01:01:01:04    *    *    *    *    *    *    *    *    *
5     A  01:01:01:05        A*01:01  A*01:01:01:05    *    *    *    *    *    *    *    *    *</code></pre>
</div>
<div id="insertion-deletion-indel-variant-representation" class="section level3">
<h3>Insertion-Deletion (Indel) Variant Representation</h3>
<p>Indel positions in these alignments are numbered sequentially,
following the first position 5’/N-terminal of the indel, with decimal
values appended to the 5’ reference position in the indel positions,
starting from “.1”, as illustrated below.</p>
<pre><code>HLAalignments$nuc$DQB1[c(1,1138:1140),c(1:4,57:63)]
     locus      allele trimmed_allele      allele_name 53 54 54.1 54.2 54.3 54.4 55
1     DQB1 05:01:01:01     DQB1*05:01 DQB1*05:01:01:01  T  C    .    .    .    .  A
1138  DQB1      06:421    DQB1*06:421      DQB1*06:421  *  *    .    .    .    .  *
1139  DQB1     06:422N   DQB1*06:422N     DQB1*06:422N  T  C    T    G    T    C  A
1140  DQB1     06:423N   DQB1*06:423N     DQB1*06:423N  *  *    .    .    .    .  *

HLAalignments$gen$DQB1[c(1,378:380),c(1:4,589:594)]
    locus      allele trimmed_allele      allele_name 52 52.1 52.2 52.3 52.4 53
1    DQB1 05:01:01:01     DQB1*05:01 DQB1*05:01:01:01  G    .    .    .    .  T
378  DQB1 06:41:01:03     DQB1*06:41 DQB1*06:41:01:03  G    .    .    .    .  T
379  DQB1     06:422N   DQB1*06:422N     DQB1*06:422N  G    T    C    T    G  T
380  DQB1      06:424    DQB1*06:424      DQB1*06:424  G    .    .    .    .  T</code></pre>
<p>Indels in <em>codon</em> alignment positions are similarly numbered
from the codon in the N-terminal direction. However, due to a
peculiarity of the R environment, which generates unique column-header
names, the last two positions of the N-terminal-ward codon
<strong>before</strong> the deletion will be numbered with the decimal
values following those used to identify the insertion positions, as
shown below. This only occurs when indel positions are included in a
selected range. In this example, the <em>allele</em> and
<em>trimmed_allele</em> columns have been omitted for spacing.</p>
<pre><code>HLAalignments$codon$DQB1[c(1,1138:1140),c(1,4,56:65)]
     locus      allele_name -15 -15.5 -15.6 -15.1 -15.2 -15.3 -15.4 -14 -14.1 -14.2
1     DQB1 DQB1*05:01:01:01   G     T     C     .     .     .     .   A     C     C
1138  DQB1      DQB1*06:421   *     *     *     .     .     .     .   *     *     *
1139  DQB1     DQB1*06:422N   G     T     C     T     G     T     C   A     C     C
1140  DQB1     DQB1*06:423N   *     *     *     .     .     .     .   *     *     *</code></pre>
<hr />
</div>
</div>
</div>
<div id="working-with-sequence-alignments" class="section level1">
<h1>2. Working with Sequence Alignments</h1>
<p>As noted above, the <strong>HLAtools</strong> package includes
several functions for working with sequence alignments. All of these
functions require that alignments first be built, as the complete set of
protein, codon, nucleotide and genomic sequence alignments is too large
to include in the <strong>HLAtools</strong> package. <span style="color:red"><strong>The functions that perform operations in
sequence alignments expect the alignments to be in a variable
named</strong> <strong><em>HLAalignments</em></strong>.</span> For
example the HLA-A protein alignment is identified as
<em>HLAalignments$prot$A</em>. If alignments are built into a different
data object, they will not be accessible to the functions described
below.</p>
<div id="building-alignments" class="section level2">
<h2>Building Alignments</h2>
<p><em>alignmentFull()</em> is a wrapper function that applies
<em>buildAlignments()</em> to populate the <em>HLAalignments</em>
object. When <em>alignmentFull()</em> is run, the resulting alignments
will be built using the IPD-IMGT/HLA Database release under which the
<em>HLAgazeteer</em> object was built (‘HLAgazeteer$version’). When
alignments for a different database release are desired,
<em>updateAll()</em> should be used to update the <em>HLAgazeteer</em>
to the desired release version before applying
<em>alignmentFull()</em>.</p>
<p><strong>Requirements</strong></p>
<p>The <em>alignmentFull()</em> function requires internet access, and
can take significant time to run to build a full complement of
alignments.</p>
<p>For example, to build all of the alignments in IPD-IMGT/HLA Database
release version 3.56.0, <em>HLAalignments &lt;- alignmentFull()</em>
completed in 6.2 minutes on a 3.3 GHz 12-Core Intel Xenon W 2019 Mac Pro
with 288 GB of 2933 MHZ DDR4 RAM, 6.5 minutes on a 2.4 GHz 8-Core Intel
Core i9 2019 MacBook Pro with 64 GB of 2667 MHZ DDR4 RAM, and 9.9
minutes on a 2.4 GHz Intel Core i9-10885H Dell Precision 5500 with 32 GB
of RAM.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>loci</em>: A vector of the locus names for which alignments
should be built. The default value (“all”) specifies all loci. Allowed
loci are listed in <em>HLAgazeteer$align</em>.</li>
<li><em>alignType</em>: A vector of alignment types for which alignments
should be built. Allowed values are “prot”, “codon”, “nuc” and “gen”,
which specify respective protein, codon, nucleotide and genomic
alignments, as well as the default value (“all”), which specifies all
four alignment types. Loci with alignments for a specific
<em>alignType</em> are defined in <em>HLAgazeteer$prot</em>,
<em>HLAgazeteer$nuc</em> and <em>HLAgazeteer$gen</em>.</li>
<li><em>version</em>: A character string identifying the IPD-IMGT/HLA
Database release version for which alignments should be built. The
default value (“Latest”) generates alignments for the most recent
IPD-IMGT/HLA Database release. Other releases are identified by their
IPD-IMGT/HLA Database version; e.g., “3.55.0” identifies IPD-IMGT/HLA
Database version 3.55.0.</li>
</ul>
<p><strong>Return</strong> A list object containing data frames of
protein (<em>prot</em>), codon (<em>codon</em>), coding nucleotide
(<em>nuc</em>), and genomic nucleotide (<em>gen</em>) alignments for
specified genes in the specified IPD-IMGT/HLA Database release, and a
character string identifying the pertinent reference database version is
returned.</p>
<p>Generate a full set of all alignments for all supported genes in the
current release.</p>
<pre><code>HLAalignments &lt;- alignmentFull()</code></pre>
<p>Generate all alignments for four genes in the current release.</p>
<pre><code>HLAalignments &lt;- alignmentFull(loci = c(&quot;C&quot;,&quot;DQB1&quot;,&quot;DPA1&quot;,&quot;DRB5&quot;)) </code></pre>
<p>Generate a protein alignment for one gene in release version
3.54.0.</p>
<pre><code>updateAll(updateType = &quot;HLAgazeteer&quot;,version = &quot;3.54.0&quot;)
HLAalignments &lt;- alignmentFull(&quot;DRB1&quot;,&quot;prot&quot;,&quot;3.54.0&quot;) </code></pre>
</div>
<div id="working-with-alignments-in-past-releases" class="section level2">
<h2>Working with Alignments in Past Releases</h2>
<p>As indicated above, <em>alignmentFull()</em> can be applied for any
IPD-IMGT/HLA Database release version. The <em>updateAll()</em> function
must be applied to build the <em>HLAgazeteer</em> for the desired
release version <strong>before</strong> applying
<em>alignmentFull()</em>.</p>
<pre><code>&gt; updateAll()
IMGTHLAGeneTypes for version 12-10-2022 is already loaded.
HLAgazeteer for version 3.56.0 has been built and loaded.
alleleListHistory for version 3.56.0 has been built and loaded
fragmentFeatureNames for version 3.56.0 has been built and loaded.
HLAatlas for version 3.56.0 has been built and loaded.

HLAalignments &lt;- alignmentFull()</code></pre>
<div id="caveats-for-alignments-from-previous-releases" class="section level3">
<h3>Caveats for Alignments from Previous Releases</h3>
<p>Prior to IPD-IMGT/HLA Database release version 3.24.0, the names of
the sequence alignment files for the HLA-DP and HLA-DQ genes in the
IPD-IMGT/HLA GitHub Repository did not include a numerical suffix in the
gene name (e.g., the protein sequence alignment file for the DQA1 gene
was ‘DQA_prot.txt’) because sequence alignment files for the DPA2, DPB2,
DQA2 and DQB2 genes had not been made available.</p>
<p>Similarly, prior to release version 3.51.0, the ‘DRB_prot.txt’ and
‘DRB_nuc.txt’ alignment files were the source of alignments for DRB1,
DRB3, DRB4 and DRB5. In release versions 3.0.0 to 3.23.0, genomic DRB1
alignments are in ‘DRB_gen.txt’ files, and these alignments are in
‘DRB1_gen.txt’ files in subsequent releases. Building DPA1, DPB1, DQA1,
DQB1, DRB1, DRB3 and DRB5 sequence alignments from these earlier
releases may require specifying a gene name that does not include the
numerical suffix. When the <em>HLAgazeteer</em> has been updated to the
desired release version, the available gene names for that release can
be found in the <em>HLAgazeteer</em>.</p>
<p>In all version 3.*.* releases (3.0.0 to 3.57.0), nucleotide
alignments for the DRB2, DRB6, DRB7 and DRB9 genes are included in the
‘DRB_nuc.txt’ file.</p>
<p>As noted above, when building alignments for past releases, the
<em>HLAgazeteer</em> should be updated to reflect the gene names
included in that release, and consulted to determine which gene names
should be provided to alignmentFull().</p>
<pre><code>updateAll(version = &quot;3.20.0&quot;)
IMGTHLAGeneTypes for version 12-10-2022 is already loaded.
HLAgazeteer for version 3.20.0 has been built and loaded.
alleleListHistory for version 3.20.0 has been built and loaded
fragmentFeatureNames for version 3.20.0 has been built and loaded.
HLAatlas for version 3.20.0 has been built and loaded.

HLAgazeteer$version
[1] &quot;3.20.0&quot;
HLAgazeteer$align
 [1] &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;DMA&quot; &quot;DMB&quot; &quot;DOA&quot; &quot;DOB&quot; &quot;DPA&quot; &quot;DPB&quot; &quot;DQA&quot; &quot;DQB&quot; &quot;DRA&quot; &quot;DRB3&quot; &quot;DRB4&quot; &quot;DRB&quot; &quot;E&quot; &quot;F&quot; &quot;G&quot; &quot;H&quot; &quot;J&quot; &quot;K&quot; &quot;L&quot; &quot;MICA&quot; &quot;MICB&quot; &quot;P&quot; &quot;TAP1&quot; &quot;TAP2&quot; &quot;V&quot; &quot;Y&quot;

HLAalignments &lt;- alignmentFull(c(&quot;C&quot;,&quot;DQB&quot;),alignType = c(&quot;nuc&quot;,&quot;gen&quot;),version = &quot;3.20.0&quot;)</code></pre>
<hr />
</div>
</div>
</div>
<div id="search-and-query-functions" class="section level1">
<h1>3. Search and Query Functions</h1>
<div id="identifying-differences-between-alleles-at-a-locus" class="section level2">
<h2>Identifying Differences Between Alleles at a Locus</h2>
<p>The <em>compareSequences()</em> function identifies sequence
differences between two alleles at a locus for a specific type of
alignment.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>alignType</em>: the alignment type of either “codon”, “gen”,
nuc” or “prot”.</li>
<li><em>alleles</em>: a vector of the two alleles to be compared, e.g.,
“c(”DPA1*01:03:38:01”,“DPA1*01:03:38:02”)“. The ‘HLA-’ prefix is
excluded for HLA alleles.</li>
</ul>
<p><strong>Return</strong> If there are no differences between the
alleles for the specified ‘alignType’, a message is returned. When there
are differences, a data frame identifying the positions and sequences
variants that distinguish the alleles is returned.</p>
<pre><code>compareSequences(&quot;prot&quot;,c(&quot;DPA1*01:03:38:01&quot;,&quot;DPA1*01:03:38:02&quot;))
[1] &quot;There are no differences between DPA1*01:03:38:01 and DPA1*01:03:38:02 in the protein alignment.&quot;

compareSequences(&quot;nuc&quot;,c(&quot;DPA1*01:03:38:01&quot;,&quot;DPA1*01:03:38:02&quot;))
[1] &quot;There are no differences between DPA1*01:03:38:01 and DPA1*01:03:38:02 in the nucleotide alignment.&quot;

compareSequences(&quot;codon&quot;,c(&quot;DPA1*01:03:38:01&quot;,&quot;DPA1*01:03:38:02&quot;))
[1] &quot;There are no differences between DPA1*01:03:38:01 and DPA1*01:03:38:02 in the codon alignment.&quot;

compareSequences(&quot;gen&quot;,c(&quot;DPA1*01:03:38:01&quot;,&quot;DPA1*01:03:38:02&quot;))
       allele_name 1544 1723 3318 4149
1 DPA1*01:03:38:01    G    G    C    G
2 DPA1*01:03:38:02    A    A    G    C</code></pre>
</div>
<div id="searching-allele-names-across-ipd-imgthla-database-release-versions" class="section level2">
<h2>Searching Allele Names Across IPD-IMGT/HLA Database Release
Versions</h2>
<p>The <em>queryRelease()</em> function searches the AlleleListHistory
object for user-defined allele name variants in a specific IPD-IMGT/HLA
release.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>rel</em>: An IPD-IMGT/HLA Database release version, represented
as either a character (e.g. “3.56.0”) or a numeric (e.g., 3560)
value.</li>
<li><em>variant</em>: A character string of any part of a locus or
allele name (e.g., “DR”, “02:01”, “DRB1*08:07”). The default (““)
specifies all alleles in ‘rel’.</li>
<li><em>all</em>: A logical. When ‘all’ = TRUE, a vector of all
instances of ‘variant’ in ‘rel’ is returned. When ‘all’ = FALSE, the
number of instances of ‘var’ in ‘rel’ is returned.</li>
</ul>
<pre><code>queryRelease(&quot;3.30.0&quot;,&quot;DRB9&quot;,FALSE) 
[1] 1
queryRelease(&quot;3.30.0&quot;,&quot;DRB9&quot;,TRUE) 
[1] &quot;DRB9*01:01&quot;
queryRelease(&quot;3.31.0&quot;,&quot;DRB9&quot;,FALSE) 
[1] 6
queryRelease(&quot;1.05.0&quot;,&quot;304&quot;,TRUE) 
[1] &quot;A*0304&quot;    &quot;A*3304&quot;    &quot;B*1304&quot;    &quot;Cw*03041&quot;  &quot;Cw*03042&quot;  &quot;DQB1*0304&quot; &quot;DRB1*0304&quot; &quot;DRB1*1304&quot; &quot;B*5304&quot;    &quot;A*2304&quot;   </code></pre>
</div>
<div id="searching-alignments-at-specific-positions" class="section level2">
<h2>Searching Alignments at Specific Positions</h2>
<p>The <em>alignmentSearch()</em> function returns the variants for
specific positions in a single allele for a specific type of
alignment.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>alignType</em>: The alignment type of either “codon”, “gen”,
nuc” or “prot”.</li>
<li><em>allelename</em>: A full-length or two-field allele name,
including the locus name, but excluding the “-HLA” prefix for HLA
alleles.</li>
<li><em>positions</em>: A vector of sequence positions (e.g.,
c(1,2,3,4,5), c(1:5), c(1,2,3,15:20,30)). If positions do not exist in
an alignment, a message is returned.</li>
<li><em>prefix</em>: A logical that indicates if the position numbers
should be included in the result (TRUE) or excluded (FALSE).</li>
<li><em>sep</em>: The value used to separate the sequences for each
position. If sep=““, a sequence block is returned.</li>
</ul>
<pre><code>alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:06:03&quot;,107)
[1] &quot;107A&quot;
alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:06:03&quot;,c(3,30,130,230))
[1] &quot;3*~30*~130G~230A&quot;
alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:06:03&quot;,c(120,125,126:128,130))
[1] &quot;120T~125C~126G~127T~128T~130G&quot;
alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:06:03&quot;,c(120,125,126:128,130),prefix=FALSE)
[1] &quot;T~C~G~T~T~G&quot;
alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:06:03&quot;,c(120,125,126:128,130),prefix=FALSE,sep=&quot;&quot;)
[1] &quot;TCGTTG&quot;
alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:06:03&quot;,c(120,125,126:128,130),prefix=FALSE,sep=&quot;:&quot;)
[1] &quot;T:C:G:T:T:G&quot;</code></pre>
<p>When sequences include indel-positions, all of the values in the
vector of sequence positions must be characters.</p>
<pre><code>alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:103&quot;,c(181:182),prefix=FALSE,sep=&quot;&quot;)
[1] &quot;GA&quot;
alignmentSearch(&quot;nuc&quot;,&quot;DPA1*01:103&quot;,c(&quot;181&quot;,&quot;181.1&quot;,&quot;181.2&quot;,&quot;182&quot;),prefix=FALSE,sep=&quot;&quot;)
[1] &quot;G..A&quot;</code></pre>
</div>
<div id="searching-alignments-for-sequence-motifs" class="section level2">
<h2>Searching Alignments for Sequence Motifs</h2>
<p>The <em>motifMatch()</em> function returns the names of alleles that
share specific nucleotide or peptide motifs.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>motif</em>: A nucleotide or peptide sequence variant motif
structured as Locus*#$~#$~#$, where # identifies a variant position, and
$ identifies the sequence variant. Any number of variants can be
specified.</li>
<li><em>alignType</em>: The alignment type of either “codon”, “gen”,
nuc” or “prot”. Only a single alignType can be specified.</li>
<li><em>full</em>: A logical value indicating if full (TRUE) or
truncated (FALSE) allele names should be returned.</li>
</ul>
<pre><code>motifMatch(&quot;A*-21M~2P&quot;,&quot;prot&quot;)
[1] &quot;A*02:774&quot;  &quot;A*11:284&quot;  &quot;A*11:417N&quot; &quot;A*68:216N&quot;

motifMatch(&quot;A*196G~301A~3046T&quot;,&quot;gen&quot;,FALSE)
[1] &quot;A*01:09&quot;

motifMatch(&quot;A*196G~301A~3046T&quot;,&quot;gen&quot;,TRUE)
[1] &quot;A*01:09:01:01&quot; &quot;A*01:09:01:02&quot;</code></pre>
</div>
<div id="building-custom-alignments" class="section level2">
<h2>Building Custom Alignments</h2>
<p>The <em>customAlign()</em> function returns a customized peptide,
codon, coding nucleotide or genomic nucleotide alignment specified
alleles at specified positions.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>alignType</em> The alignment type of either “codon”, “gen”, nuc”
or “prot”.</li>
<li><em>alleles</em> A vector of allele names, which can incude alleles
at multiple loci. The ‘HLA-’ prefix is excluded for HLA alleles.</li>
<li><em>positions</em> Either a single vector positions, against which
all loci will be aligned, or a list of vectors of positions, for which
there is one vector for each allele in ‘alleles’. Each allele in
‘alleles’ will be aligned on the basis of the corresponding vector.</li>
</ul>
<pre><code>customAlign(&quot;codon&quot;,c(&quot;DPB1*01:01:01:01&quot;,&quot;DQA1*01:01:01:01&quot;,&quot;DQB1*05:01:01:01&quot;),c(1:4,6:9))
            Allele   1   2   3   4   6   7   8   9
1 DPB1*01:01:01:01 AGG GCC ACT CCA AAT TAC GTG TAC
2 DQA1*01:01:01:01 GAA GAC ATT GTG GAC CAC GTT GCC
3 DQB1*05:01:01:01 AGA GAC TCT CCC GAT TTC GTG TAC</code></pre>
<p>When the length of vectors in a ‘positions’ list varies, it is
recommended to coordinate the values in ‘alleles’ and ‘positions’ so
that the longest vector is generated first.</p>
<pre><code>customAlign(&quot;codon&quot;,c(&quot;DPB1*01:01:01:01&quot;,&quot;DQA1*01:01:01:01&quot;,&quot;DQB1*05:01:01:01&quot;),list(19:35,1:4,6:9))
              DPB1  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
1 DPB1*01:01:01:01 AAT GGG ACA CAG CGC TTC CTG GAG AGA TAC ATC TAC AAC CGG GAG GAG TAC
2             DQA1   1   2   3   4                                                    
3 DQA1*01:01:01:01 GAA GAC ATT GTG                                                    
4             DQB1   6   7   8   9                                                    
5 DQB1*05:01:01:01 GAT TTC GTG TAC</code></pre>
<hr />
</div>
</div>
<div id="data-translation-and-conversion-tools" class="section level1">
<h1>4. Data Translation and Conversion Tools</h1>
<div id="working-with-genotype-list-string-gl-string-gl-string-code-glsc-and-uniformat-data" class="section level2">
<h2>Working with Genotype List String (GL String), GL String Code
(GLSC), and UNIFORMAT Data</h2>
<p><a href="https://onlinelibrary.wiley.com/doi/10.1111/tan.15126">Genotype
List (GL) String</a> is a grammar for describing the relationships
between allele names reported in a genotype, using a set of six
case-defined “AND” and “OR” delimiters (‘?’, ‘^’, ‘|’, ‘+’, ‘~’, ‘~’,
and ‘/’).</p>
<p><a href="https://hla-net.eu/tools/overview/">UNIFORMAT</a> is a
parallel grammar that uses an alternate set of operators (“,”, “|”, ” “,
and”~“) to describe the relationships between allele names in a
genotype. Additional details of UNIFORMAT can be found at <a href="https://hla-net.eu/tools/uniformate/">hla-net.eu</a>.</p>
<p><a href="https://onlinelibrary.wiley.com/doi/10.1111/tan.15145">GL
String Code</a> (GLSC) is a format that encapsulates a GL String with
information describing the associated gene family namespace and the
nomenclature version (or date) under which the GL String was generated.
Additional information about GL String and GL String Code can be found
at <a href="https://glstring.org">glstring.org</a>.</p>
<p>An example GLSC for an HLA GL String generated under IPD-IMGT/HLA
Database release version 3.01.0 is presented below.</p>
<pre><code>hla#3.1.0#HLA-A*02:01~HLA-C*07:02~HLA-B*07:02~HLA-DRB1*15:01~HLA-DQB1*06:02~HLA-DPB1*04:02+HLA-A*01:01~HLA-C*06:02~HLA-B*57:01~HLA-DRB1*07:01~HLA-DQB1*03:03~HLA-DPB1*04:01&quot;</code></pre>
</div>
<div id="translate-gl-string-codes-across-ipd-imgthla-database-release-versions" class="section level2">
<h2>Translate GL String Codes Across IPD-IMGT/HLA Database Release
Versions</h2>
<div id="updategl" class="section level3">
<h3>updateGL</h3>
<p>The <em>updateGL()</em> function translates the HLA allele names in a
GL String Code across IPD-IMGT/HLA Database release versions, going back
to version 1.05.0. Allele names can be matched across release versions
via the allele Accession ID, and via allele name fields shared across
alleles in a given release, as illustrated below.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>GLStringCode</em> A string of HLA allele names and operators in
GL String Code format, signifying their relation with one another and
the associated IMGT/HLA Database release version.</li>
<li><em>Version</em> A string identifying of the desired IPD-IMGT/HLA
Database release version to which the alleles should be updated.</li>
<li><em>expand</em> A logical that indicates whether user would like to
return all allele names that match the allele name in ‘GLStringCode’
(TRUE), or if only the direct HLA accession ID match should be returned
(FALSE).</li>
</ul>
<p>In the examples below, a GLSC containing a single DPB1 allele name,
as recorded in release version 1.05.0, is translated to version 3.52.0.
In the first example, ‘expand = FALSE’, and the HLA accession ID of the
submitted allele is used to perform the translation to a single allele.
In the second example, ‘expand = TRUE’ and all alleles that match the
submitted allele in release 3.52.0 are returned.</p>
<pre><code>updateGL(&quot;hla#1.05.0#HLA-DPA1*0106&quot;, &quot;3.52.0&quot;, expand = FALSE)
[1] &quot;hla#3.52.0#HLA-DPA1*01:06:01&quot;

updateGL(&quot;hla#1.05.0#HLA-DPA1*0106&quot;, &quot;3.52.0&quot;, expand = TRUE)
[1] &quot;hla#3.52.0#HLA-DPA1*01:06:01/HLA-DPA1*01:06:02/HLA-DPA1*01:06:03&quot;</code></pre>
</div>
<div id="multiupdategl" class="section level3">
<h3>multiUpdateGL</h3>
<p>The <em>multiUpdateGL()</em> function translates columns of GL String
Code data across IPD-IMGT/HLA Database release versions in data frames.
If a single column of a data frame is provided, it must contain GL
string codes. If multiple columns are entered, the first column should
contain identifiers, which will not be updated. Translations can be
performed for all IPD-IMGT/HLA Database release versions in the loaded
<em>alleleListHistory</em> data object, going back to IPD-IMGT/HLA
Database version 1.05.0.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>GLstringArray</em> An array of HLA allele names and operators in
GL String code format identifying their relation with one another and
the pertinent IPD-IMGT/HLA Database release version.</li>
<li><em>Version</em> A text string identifying the desired version to
which for the alleles should be updated.</li>
<li><em>expand</em> A logical specifying whether to include only the
direct HLA ID match (FALSE), or all possible allele matches (TRUE). The
default value is FALSE.</li>
</ul>
<pre><code>GLstringArray
multiUpdateGL(GLSC.ex[[2]][1], Version = &quot;3.53.0&quot;)
1 hla#3.53.0#HLA-A*02:01:01:01~HLA-C*07:02:01:01~HLA-B*07:02:01:01~HLA-DRB1*15:01:01:01~HLA-DQB1*06:02:01:01~HLA-DPB1*04:02:01:01+HLA-A*01:01:01:01~HLA-C*06:02:01:01~HLA-B*57:01:01:01~HLA-DRB1*07:01:01:01~HLA-DQB1*03:03:02:01~HLA-DPB1*04:01:01:01</code></pre>
</div>
</div>
<div id="translate-vectors-and-data-frames-of-hla-allele-name-data-across-ipd-imgthla-database-releases" class="section level2">
<h2>Translate vectors and data frames of HLA allele name data across
IPD-IMGT/HLA Database releases</h2>
<p>The GIANT() function applies GLupdate() to translate HLA allele name
data across IPD-IMGT/HLA Database release versions, back to version
1.05.0, and functions on both vectors and data frames of HLA allele name
data, including BIGDAWG-formatted HLA datasets. Data frames that are not
BIGDAWG-formatted must be entirely composed of columns of HLA allele
name data.</p>
<p><em>Translation of vectors of HLA allele names</em></p>
<pre><code>GIANT(c(&quot;A*01:01:01:01&quot;,&quot;DQA1*01:01:01:01&quot;),&quot;3.56.0&quot;,&quot;2.20.0&quot;)
[1] &quot;A*01010101&quot;  &quot;DQA1*010101&quot;
GIANT(c(&quot;A*01:01:01:01&quot;,&quot;DQA1*01:01:01:01&quot;),&quot;3.56.0&quot;,&quot;3.00.0&quot;)
[1] &quot;A*01:01:01:01&quot; &quot;DQA1*01:01:01&quot;
GIANT(c(&quot;A*01:01:01:01&quot;,&quot;DQA1*01:01:01:01&quot;),&quot;3.56.0&quot;,&quot;1.09.0&quot;)
[1] &quot;A*01011&quot;   &quot;DQA1*0101&quot;</code></pre>
<p><em>Translation of BIGDAWG-formatted HLA data</em></p>
<pre><code>sHLAdata[1:5,c(1:8,15:18)]
   Subject Status           A         A.1           C         C.1           B         B.1
1 UT900-23      0        &lt;NA&gt;        &lt;NA&gt; 01:02:01:01    02:10:06 13:01:01:01    18:01:02
2 UT900-24      0 01:01:01:01 02:01:01:01 03:07:01:01       06:05 14:01:01:01 39:02:01:01
3 UT900-25      0       02:10    03:01:02       07:12 01:02:01:01       15:20 13:01:01:01
4 UT900-26      0 01:01:01:01       02:18 08:04:01:01    12:02:01 35:09:01:01 40:05:01:01
5 UT910-01      0 25:01:01:01 02:01:01:01 15:07:01:01 03:07:01:01    51:01:03 14:01:01:01

         DPA1      DPA1.1        DPB1      DPB1.1
1 02:01:01:01 02:01:01:01      103:01 14:01:01:01
2 01:03:01:01    02:01:03 14:01:01:01 14:01:01:01
3 01:03:01:01 01:03:01:01      103:01       58:01
4 01:03:01:01 01:03:01:01 14:01:01:01      103:01
5 01:03:01:01 02:01:01:01 14:01:01:01       58:01

GIANT(sHLAdata,&quot;3.56.0&quot;,&quot;2.10.0&quot;)[1:5,c(1:8,15:18)]
   Subject Status        A      A.1      C    C.1      B    B.1   DPA1 DPA1.1 DPB1 DPB1.1
1 UT900-23      0     &lt;NA&gt;     &lt;NA&gt; 010201 020205   1301 180102 020101 020101 0403   1401
2 UT900-24      0 01010101 02010101   0307   0605   1401 390201 010301 020103 1401   1401
3 UT900-25      0     0210   030102   0712 010201   1520   1301 010301 010301 0403   5801
4 UT900-26      0 01010101     0218   0804 120201 350901   4005 010301 010301 1401   0403
5 UT910-01      0   250101 02010101   1507   0307 510103   1401 010301 020101 1401   5801

GIANT(sHLAdata,&quot;3.56.0&quot;,&quot;1.15.0&quot;)[1:5,c(1:8,15:18)]
DPB1*103:01 has been removed from the Allele List or has had its name changed.
DPB1*103:01 has been removed from the Allele List or has had its name changed.
   Subject Status     A   A.1    C   C.1     B   B.1  DPA1 DPA1.1 DPB1 DPB1.1
1 UT900-23      0  &lt;NA&gt;  &lt;NA&gt; 0102 02025  1301 18012 02011  02011 &lt;NA&gt;   1401
2 UT900-24      0 01011 02011 0307  0605  1401 39021 01031  02013 1401   1401
3 UT900-25      0  0210 03012 0712  0102  1520  1301 01031  01031 &lt;NA&gt;   5801
4 UT900-26      0 01011  0218 0804 12021 35091  4005 01031  01031 1401   &lt;NA&gt;
5 UT910-01      0  2501 02011 1507  0307 51013  1401 01031  02011 1401   5801</code></pre>
<p>In these examples, the DPB1*103:01 allele was originally named
DPB1*04:03, but this allele was not identified until release version
2.0.0.</p>
</div>
<div id="convert-gl-string-formatted-data-to-uniformat-formatted-data" class="section level2">
<h2>Convert GL String-formatted data to UNIFORMAT-formatted data</h2>
<div id="validateglstring" class="section level3">
<h3>validateGLstring</h3>
<p>The <em>validateGLstring()</em> function evaluates version 1.0 and
1.1 GL Strings for incorrect characters. TRUE is returned when all
characters in the GL String are valid for the specified version. FALSE
is returned when the GL String contains invalid characters for the
version.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>GLstring</em> A character string of GL String-formatted HLA
allele names and GL String 1.0 or 1.1 operators.</li>
<li><em>version</em> A character string of either ‘1.0’ or ‘1.1’
identifying the GL String version.</li>
</ul>
<pre><code>GLstring &lt;- &quot;HLA-A*02:01/HLA-A*02:02?HLA-A*03:01/HLA-A*03:02&quot;
validateGLstring(GLstring,&quot;1.0&quot;)
HLA-A*02:01/HLA-A*02:02?HLA-A*03:01/HLA-A*03:02 contains characters or operators not permitted in GL String version 1.0.
[1] FALSE
validateGLstring(GLstring,&quot;1.1&quot;)
[1] TRUE</code></pre>
</div>
<div id="glstouni" class="section level3">
<h3>GLStoUNI</h3>
<p>The <em>GLStoUNI()</em> function converts genotype data from GL
String format to UNIFORMAT format. Because the <a href="https://glstring.org/syntax-1.1.html">GL String ‘?’ delimeter</a>
has no UNIFORMAT cognate, this function does not support the ‘?’
delimiter or the <a href="https://pubmed.ncbi.nlm.nih.gov/37286192">GL
String v1.1</a> format.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>GLstring</em> A character string of GL String-formatted HLA
allele names and GL String 1.0 operators.</li>
<li><em>prefix</em> A string containing a gene-system prefix (default is
“HLA-”).</li>
<li><em>pre</em> A logical that indicates whether ‘prefix’ should be
appended to the allele names (TRUE) or not (FALSE). The default value is
FALSE.</li>
</ul>
<pre><code>GLStoUNI(&quot;HLA-A*02:01/HLA-A*02:02+HLA-A*03:01/HLA-A*03:02&quot;)
[1] &quot;A*02:01,A*03:01|A*02:01,A*03:02|A*02:02,A*03:01|A*02:02,A*03:02&quot;</code></pre>
</div>
<div id="multiglstouni" class="section level3">
<h3>multiGLStoUNI</h3>
<p>The <em>multiGLStoUNI()</em> function converts GL string formatted
data in data frames or vectors to uniformat format. If a vector is
provided, all elements must be GL Strings. If a data frame is provided,
the first column is assumed to contain identifiers and will not be
converted; the remaining columns should contain GL strings.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>GLstringArray</em> A data frame or a vector of GL String 1.0
formatted data. If ‘GLstringArray’ is a data frame with more than one
column, the first column should contain only identifiers. If
‘GLstringArray’ is a vector, it should contain only GL Strings.</li>
<li><em>prefix</em> A string containing a gene-system prefix (default is
“HLA-”).</li>
<li><em>pre</em> A logical that indicates whether ‘prefix’ should be
appended to the allele names (TRUE) or not (FALSE). The default value is
FALSE.</li>
</ul>
<p><strong>Return</strong> If ‘GLstringArray’ is a data frame, a data
frame of UNIFORMAT values is returned. If ‘GLstringArray’ is a vector, a
vector of UNIFORMAT values is returned.</p>
<p><em>Example GL String data frame</em></p>
<pre><code>GLstring.ex[1:3,]
Relation    Gl.String
Subject    1 HLA-A*02:01~HLA-C*07:02~HLA-B*07:02~HLA-DRB1*15:01~HLA-DQB1*06:02~HLA-DPB1*04:02+HLA-A*01:01~HLA-C*06:02~HLA-B*57:01~HLA-DRB1*07:01~HLA-DQB1*03:03~HLA-DPB1*04:01
Subject    2 HLA-A*03:01~HLA-C*07:01~HLA-B*49:01~HLA-DRB1*04:05~HLA-DQB1*03:02~HLA-DPB1*02:01+HLA-A*01:01~HLA-C*07:01~HLA-B*08:01~HLA-DRB1*13:01~HLA-DQB1*06:03~HLA-DPB1*04:01
Subject    3 HLA-A*11:01~HLA-C*04:01~HLA-B*15:01~HLA-DRB1*04:01~HLA-DQB1*03:02~HLA-DPB1*04:02+HLA-A*03:01~HLA-C*08:02~HLA</code></pre>
<p><em>Example GL String vectors</em></p>
<pre><code>GLstring.ex[2][1:3,]
[1] &quot;HLA-A*02:01~HLA-C*07:02~HLA-B*07:02~HLA-DRB1*15:01~HLA-DQB1*06:02~HLA-DPB1*04:02+HLA-A*01:01~HLA-C*06:02~HLA-B*57:01~HLA-DRB1*07:01~HLA-DQB1*03:03~HLA-DPB1*04:01&quot;
[2] &quot;HLA-A*03:01~HLA-C*07:01~HLA-B*49:01~HLA-DRB1*04:05~HLA-DQB1*03:02~HLA-DPB1*02:01+HLA-A*01:01~HLA-C*07:01~HLA-B*08:01~HLA-DRB1*13:01~HLA-DQB1*06:03~HLA-DPB1*04:01&quot;
[3] &quot;HLA-A*11:01~HLA-C*04:01~HLA-B*15:01~HLA-DRB1*04:01~HLA-DQB1*03:02~HLA-DPB1*04:02+HLA-A*03:01~HLA-C*08:02~HLA-B*14:02~HLA-DRB1*13:02~HLA-DQB1*06:09~HLA-DPB1*04:01&quot;</code></pre>
<p><em>Convert GL String to UNIFORMAT</em></p>
<pre><code>multiGLStoUNI(GLstring.ex[1:5,1:2])
  Relation                                                                                                         Gl.String
1  Subject A*02:01~C*07:02~B*07:02~DRB1*15:01~DQB1*06:02~DPB1*04:02,A*01:01~C*06:02~B*57:01~DRB1*07:01
~DQB1*03:03~DPB1*04:01
2  Subject A*03:01~C*07:01~B*49:01~DRB1*04:05~DQB1*03:02~DPB1*02:01,A*01:01~C*07:01~B*08:01~DRB1*13:01
~DQB1*06:03~DPB1*04:01
3  Subject A*11:01~C*04:01~B*15:01~DRB1*04:01~DQB1*03:02~DPB1*04:02,A*03:01~C*08:02~B*14:02~DRB1*13:02
~DQB1*06:09~DPB1*04:01
4  Subject A*68:01~C*15:02~B*40:06~DRB1*16:02~DQB1*05:02~DPB1*10:01,A*68:01~C*06:02~B*45:01~DRB1*04:05
~DQB1*03:02~DPB1*11:01
5  Subject A*02:01~C*05:01~B*44:02~DRB1*15:01~DQB1*06:02~DPB1*02:01,A*30:01~C*06:02~B*13:02~DRB1*07:01
~DQB1*02:02~DPB1*17:01

multiGLStoUNI(GLstring.ex[[2]][1:5])
[1] &quot;A*02:01~C*07:02~B*07:02~DRB1*15:01~DQB1*06:02~DPB1*04:02,A*01:01~C*06:02~B*57:01~
DRB1*07:01~DQB1*03:03~DPB1*04:01&quot;
[2] &quot;A*03:01~C*07:01~B*49:01~DRB1*04:05~DQB1*03:02~DPB1*02:01,A*01:01~C*07:01~B*08:01~
DRB1*13:01~DQB1*06:03~DPB1*04:01&quot;
[3] &quot;A*11:01~C*04:01~B*15:01~DRB1*04:01~DQB1*03:02~DPB1*04:02,A*03:01~C*08:02~B*14:02~
DRB1*13:02~DQB1*06:09~DPB1*04:01&quot;
[4] &quot;A*68:01~C*15:02~B*40:06~DRB1*16:02~DQB1*05:02~DPB1*10:01,A*68:01~C*06:02~B*45:01~
DRB1*04:05~DQB1*03:02~DPB1*11:01&quot;
[5] &quot;A*02:01~C*05:01~B*44:02~DRB1*15:01~DQB1*06:02~DPB1*02:01,A*30:01~C*06:02~B*13:02~
DRB1*07:01~DQB1*02:02~DPB1*17:01&quot;
</code></pre>
</div>
</div>
<div id="convert-uniformat-formatted-data-to-gl-string-formatted-data" class="section level2">
<h2>Convert UNIFORMAT-formatted data to GL String-formatted data</h2>
<div id="validateuniformat" class="section level3">
<h3>validateUniformat</h3>
<p>The <em>validateUniformat()</em> function evaluates UNIFORMAT data
for incorrect characters. TRUE is returned when all characters in the
UNIFORMAT data are valid for the specified version. FALSE is returned
when the UNIFORMAT data contain invalid characters.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>uniformat</em> A character string of allele names and operators
in the UNIFORMAT format.</li>
</ul>
<pre><code>validateUniformat(&quot;A*02:01,A*03:01|A*02:01,A*03:02|A*02:02,A*03:01|A*02:02,A*03:02&quot;)
[1] TRUE

validateUniformat(&quot;A*02:01+A*03:01|A*02:01,A*03:02|A*02:02,A*03:01|A*02:02,A*03:02&quot;)
A*02:01+A*03:01|A*02:01,A*03:02|A*02:02,A*03:01|A*02:02,A*03:02 contains operators or characters not permitted in UNIFORMAT.
[1] FALSE</code></pre>
</div>
<div id="unitogls" class="section level3">
<h3>UNItoGLS</h3>
<p>The <em>UNItoGLS()</em> function converts a single UNIFORMAT string
to GL string format.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>uniformat</em> A string of HLA allele names and operators in the
UNIFORMAT format signifying their relation with one another.</li>
<li><em>prefix</em> A character string of the desired gene-system prefix
(default is “HLA-”).</li>
<li><em>pre</em> A logical that indicates whether returned allele names
should contain ‘prefix’ (TRUE), or if ‘prefix’ should be excluded from
returned names (FALSE).</li>
</ul>
<pre><code>UNItoGLS(&quot;A*02:01,A*03:01|A*02:01,A*03:02|A*02:02,A*03:01|A*02:02,A*03:02&quot;)
[1] &quot;HLA-A*02:01/HLA-A*02:02+HLA-A*03:01/HLA-A*03:02&quot;</code></pre>
</div>
<div id="multiunitogls" class="section level3">
<h3>multiUNItoGLS</h3>
<p>The <em>multiUNItoGLS()</em> function translates columns of arrays
from uniformat format to GL string format. If one column is entered, it
must be in UNIFORMAT format. If multiple columns are entered, the first
column (furthest left) should contain an identification factor of some
kind, as the first column will not be translated, and the remaining
columns should contain UNIFORMAT strings. Examples of input formats are
presented below.</p>
<p><em>Example UNIFORMAT Data Frame</em></p>
<pre><code>&gt; UNIFORMAT.example[1:3,]
   sample.id  genotype
1 hhrv_id195  blank,A*02:01|A*02:01,A*02:01 blank,B*44:03|B*44:03,B*44:03 blank,DRB1*07:01|DRB1*07:01,DRB1*07:01
2 hhrv_id454  A*02:01,A*23:01|A*02:01,A*23:17 B*27:05,B*44:03|B*27:13,B*44:03 DRB1*07:01,DRB1*11:04
3 hhrv_id642  A*66:01,A*31:08|A*31:08,A*66:08 B*27:05,B*07:02|B*27:05,B*07:61|B*27:13,B*07:02|B*27:13,B*07:61 DRB1*01:01,DRB1*16:02</code></pre>
<p><em>Example UNIFORMAT Vector</em></p>
<pre><code>&gt; UNIFORMAT.example[2][1:3,]
[1] &quot;blank,A*02:01|A*02:01,A*02:01 blank,B*44:03|B*44:03,B*44:03 blank,DRB1*07:01|DRB1*07:01,DRB1*07:01&quot;
[2] &quot;A*02:01,A*23:01|A*02:01,A*23:17 B*27:05,B*44:03|B*27:13,B*44:03 DRB1*07:01,DRB1*11:04&quot;
[3] &quot;A*66:01,A*31:08|A*31:08,A*66:08 B*27:05,B*07:02|B*27:05,B*07:61|B*27:13,B*07:02|B*27:13,B*07:61 DRB1*01:01,DRB1*16:02&quot;</code></pre>
<p><strong>Parameters</strong></p>
<ul>
<li><em>UniformatArray</em> A data frame or vector of UNIFORMAT
formatted strings.</li>
<li><em>prefix</em> A character string of the desired gene-system prefix
(default is “HLA-”).</li>
<li><em>pre</em> A logical that indicates whether returned allele names
should contain ‘prefix’ (TRUE), or if ‘prefix’ should be excluded from
returned names (FALSE).</li>
</ul>
<p><strong>Return</strong> If ‘uniformat’is a data frame, a data frame
of GL String values is returned. If ’uniformat’ is a vector, a vector of
GL String values is returned.</p>
<p><em>Convert UNIFORMAT to GL String</em></p>
<pre><code>multiUNItoGLS(UNIFORMAT.example[1:3,])
   sample.id                                                                                                          genotype
1 hhrv_id195       HLA-blank/HLA-A*02:01+HLA-A*02:01^HLA-blank/HLA-B*44:03+HLA-B*44:03^HLA-blank/HLA-DRB1*07:01+HLA-DRB1*07:01
2 hhrv_id454             HLA-A*02:01+HLA-A*23:01/HLA-A*23:17^HLA-B*27:05/HLA-B*27:13+HLA-B*44:03^HLA-DRB1*07:01+HLA-DRB1*11:04
3 hhrv_id642 HLA-A*66:01/HLA-A*66:08+HLA-A*31:08^HLA-B*27:05/HLA-B*27:13+HLA-B*07:02/HLA-B*07:61^HLA-DRB1*01:01+HLA-DRB1*16:02

multiUNItoGLS(UNIFORMAT.example[2][1:3,])
[1] &quot;HLA-blank/HLA-A*02:01+HLA-A*02:01^HLA-blank/HLA-B*44:03+HLA-B*44:03^HLA-blank/HLA-DRB1*07:01+HLA-DRB1*07:01&quot;      
[2] &quot;HLA-A*02:01+HLA-A*23:01/HLA-A*23:17^HLA-B*27:05/HLA-B*27:13+HLA-B*44:03^HLA-DRB1*07:01+HLA-DRB1*11:04&quot;            
[3] &quot;HLA-A*66:01/HLA-A*66:08+HLA-A*31:08^HLA-B*27:05/HLA-B*27:13+HLA-B*07:02/HLA-B*07:61^HLA-DRB1*01:01+HLA-DRB1*16:02&quot;
</code></pre>
<p><em>Convert GL String to UNIFORMAT</em></p>
<pre><code>multiGLStoUNI(GLstring.ex[1:5,1:2])
  Relation                                                                                                         Gl.String
1  Subject A*02:01~C*07:02~B*07:02~DRB1*15:01~DQB1*06:02~DPB1*04:02,A*01:01~C*06:02~B*57:01~DRB1*07:01
~DQB1*03:03~DPB1*04:01
2  Subject A*03:01~C*07:01~B*49:01~DRB1*04:05~DQB1*03:02~DPB1*02:01,A*01:01~C*07:01~B*08:01~DRB1*13:01
~DQB1*06:03~DPB1*04:01
3  Subject A*11:01~C*04:01~B*15:01~DRB1*04:01~DQB1*03:02~DPB1*04:02,A*03:01~C*08:02~B*14:02~DRB1*13:02
~DQB1*06:09~DPB1*04:01
4  Subject A*68:01~C*15:02~B*40:06~DRB1*16:02~DQB1*05:02~DPB1*10:01,A*68:01~C*06:02~B*45:01~DRB1*04:05
~DQB1*03:02~DPB1*11:01
5  Subject A*02:01~C*05:01~B*44:02~DRB1*15:01~DQB1*06:02~DPB1*02:01,A*30:01~C*06:02~B*13:02~DRB1*07:01
~DQB1*02:02~DPB1*17:01

multiGLStoUNI(GLstring.ex[[2]][1:5])
[1] &quot;A*02:01~C*07:02~B*07:02~DRB1*15:01~DQB1*06:02~DPB1*04:02,A*01:01~C*06:02~B*57:01~
DRB1*07:01~DQB1*03:03~DPB1*04:01&quot;
[2] &quot;A*03:01~C*07:01~B*49:01~DRB1*04:05~DQB1*03:02~DPB1*02:01,A*01:01~C*07:01~B*08:01~
DRB1*13:01~DQB1*06:03~DPB1*04:01&quot;
[3] &quot;A*11:01~C*04:01~B*15:01~DRB1*04:01~DQB1*03:02~DPB1*04:02,A*03:01~C*08:02~B*14:02~
DRB1*13:02~DQB1*06:09~DPB1*04:01&quot;
[4] &quot;A*68:01~C*15:02~B*40:06~DRB1*16:02~DQB1*05:02~DPB1*10:01,A*68:01~C*06:02~B*45:01~
DRB1*04:05~DQB1*03:02~DPB1*11:01&quot;
[5] &quot;A*02:01~C*05:01~B*44:02~DRB1*15:01~DQB1*06:02~DPB1*02:01,A*30:01~C*06:02~B*13:02~
DRB1*07:01~DQB1*02:02~DPB1*17:01&quot;</code></pre>
<hr />
</div>
</div>
</div>
<div id="data-analysis-tools" class="section level1">
<h1>5. Data Analysis Tools</h1>
<p>The functions described in this section either perform data analyses
or generate specific data sets suitable for additional analyses. All
consume BIGDAWG-formatted datasets.</p>
<div id="example-data" class="section level3">
<h3>Example Data</h3>
<p>To facilitate testing and experimentation with these functions, the
‘sHLAdata’ data object is included in the package. This
BIGDAWG-formatted HLA genotype data object represents completely
synthetic IPD-IMGT/HLA release 3.56.0 HLA-A, -C, -B, -DRB1, -DQA1,
-DQB1, -DPA1, and -DPB1 genotype data for 24 control subjects and 23
case subjects, and does not represent any true human population.</p>
</div>
<div id="calculating-relative-risk-with-bigdawg-formatted-datasets" class="section level2">
<h2>Calculating Relative Risk with BIGDAWG-formatted Datasets</h2>
<div id="relrisk" class="section level3">
<h3>relRisk</h3>
<p>The relRisk() function calculates relative risk (RR) for individual
alleles and genotypes in <a href="https://CRAN.R-project.org/package=BIGDAWG/vignettes/BIGDAWG.html#input-data">BIGDAWG-formatted
datasets</a>. RR analysis is intended for <em>non-case-control</em>
datasets. While two subject categories (0 and 1) are required in the
<strong>Subject</strong> column of the BIGDAWG-formatted datafile, the
categories should <em>not</em> be patients and controls. Instead, the
categories may be, e.g., two groups of patients with the same disease,
where one group has more severe symptoms (coded as 1), and other has
mild symptoms (coded as 0). In this case, relRisk() identifies the risk
of severity associated with a given allele or genotype.</p>
<p>relrisk() returns a list object of two lists (“alleles” and
“genotypes”), each of which contains relative risk, confidence interval
and p-value data for the individual alleles and individual genotypes at
each locus in a BIGDAWG-formatted non-case-control genotype data frame
or tab-delimited text file.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>dataset</em> The name of a non-case-control genotype dataset
using the BIGDAWG format.</li>
<li><em>return</em> A logical identifying if the list object should be
returned (return=TRUE), or if pairs of tab-delimited text files of
results (one for alleles and one for genotypes) should be written to the
working directory for each locus.</li>
<li><em>save.path</em> A character string identifying the path in which
to write the pair of files when return is FALSE. The default value is
tempdir().</li>
</ul>
<pre><code>rr &lt;- relRisk(sHLAdata[,1:4])</code></pre>
<p>Column headers in each returned data frame are, Locus, Variant,
Status_1, Status_0, RelativeRisk, CI.low, CI.high, p.value, and
Significant.</p>
</div>
</div>
<div id="division-of-bigdawg-formatted-datasets-for-stratification-analyses" class="section level2">
<h2>Division of BIGDAWG-formatted Datasets for Stratification
Analyses</h2>
<div id="bdstrat-multilocus-allele-stratificaiton-for-bigdawg-datasets" class="section level3">
<h3>BDstrat: Multilocus Allele Stratificaiton for BIGDAWG Datasets</h3>
<p>The BDstrat() function divides a <a href="https://CRAN.R-project.org/package=BIGDAWG/vignettes/BIGDAWG.html#input-data">BIGDAWG-formatted</a>]
dataset into two strata, one of which contains all subjects with any of
a set of specified alleles, while the other contains all subjects
without any of the specified alleles. If the positive-stratum and
negative-stratum datasets have sufficient numbers of case and control
(or exposed and unexposed) subjects, each stratum can be analyzed via
the BIGDAWG package or relRisk(). Sample size is an important
consideration when determining how many alleles on which to
stratify.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><em>dataset</em>: A BIGDAWG-formatted data frame or a path to a
BIGDAWG-formatted, tab-delimited text file.</li>
<li><em>alleles</em>: A vector of allele names in the
<locus>*<allele name> format (e.g., “A*01:01:01:01” for HLA alleles or
“KIR2DL1*001” for KIR alleles).</li>
<li><em>warnBelow</em> : An integer that defines a low number of
subjects in a stratum, generating a warning message. The default value
is 21.</li>
</ul>
<p><strong>Return</strong> A list-object of two BIGDAWG-formatted data
frames titled ‘dataset’$‘alleles’-positive’ and
‘dataset’$‘alleles’-negative`. The positive list element includes all
subjects with the specified alleles, and the negative list element
includes all subjects without those specified alleles.</p>
<pre><code>Stratify on a single allele in the HLA_data dataset bundled with BIGDAWG
HLA_data.single.strat &lt;- BDstrat(sHLAdata,&quot;DRB1*08:02:01:01&quot;) 

Stratify on three alleles at two loci in the HLA_data dataset bundled with BIGDAWG
HLA_data.multi.strat &lt;- BDstrat(sHLAdata,c(&quot;DRB1*04:02:01&quot;,&quot;DRB1*04:07:01:01&quot;,&quot;A*32:04&quot;))

Run BIGDAWG() on both multi-allele strata
for(i in 1:2) {BIGDAWG(HLA_data.multi.strat[[i]],HLA = TRUE,Run.Tests = &quot;L&quot;)}
</code></pre>
</div>
</div>
<div id="conversion-of-bigdawg-formatted-datasets-to-python-for-population-genetics-pypop-formatted-datasets" class="section level2">
<h2>Conversion of BIGDAWG-formatted Datasets to <em>Python for
Population Genetics</em> (PyPop)-formatted Datasets</h2>
<div id="bdtopypop-enabling-population-genetic-analyses-of-case-and-control-datasets" class="section level4">
<h4>BDtoPyPop : Enabling Population Genetic Analyses of Case and Control
Datasets</h4>
<ul>
<li>BDtoPyPop() converts a <a href="https://CRAN.R-project.org/package=BIGDAWG/vignettes/BIGDAWG.html#input-data">BIGDAWG-formatted</a>
case-control dataset into two <a href="http://pypop.org/docs/guide-chapter-usage.html">PyPop version
1.*.*</a> datasets – one for all ‘case’ subjects and one for all
‘control’ subjects. PyPop can be exported as a pair of “.pop” files for
analysis, or can be generated as a list object containing two data
frames for additional manipulation before analysis.</li>
</ul>
<p><strong>Parameters</strong></p>
<ul>
<li><em>dataset</em> A BIGDAWG-formatted data frame containing a
case-control dataset.</li>
<li><em>filename</em> A character string identifying either the desired
path and name for the generated files, or the elements of the returned
list object.</li>
<li><em>save.file</em> A logical indicating if the pair of files should
be written (TRUE) or if a list object should be returned. The default
value is TRUE.</li>
<li><em>save.path</em> A character string identifying the path in which
to write the pair of files when return is FALSE. The default value is
tempdir().</li>
</ul>
<p><strong>Return</strong> When save.file = TRUE a pair of
PyPop-formatted text files named “‘filename’.positive.pop” and
“‘filename’.negative.pop” are generated in the specified directory. The
“‘filename’.positive.pop” file includes all subjects with a value of 1
in the second column of the BIGDAWG data file, and the
“‘filename’.negative.pop” file includes all subjects with a value of 0
in the second column of the BIGDAWG data file.</p>
<p>When save.file = FALSE, a list-object of two PyPop-formatted data
frames titled “‘filename’.positive” and “‘filename’-negative” is
returned. The positive list element includes all subjects a value of 1
in the second column of the BIGDAWG data file, and the negative list
element includes all subjects with a value of 0 in the second column of
the BIGDAWG data file.</p>
<p><strong>Note</strong> PyPop is not an R package. In order to analyze
the files generated by this function with PyPop version 1.*.*, <a href="http://pypop.org/docs/guide-chapter-install.html">the
pypop-genomics package must be installed</a> on your system, and a <a href="http://pypop.org/docs/guide-chapter-usage.html#the-configuration-file">PyPop
configuration file</a>, detailing the analyses to be performed, is
required. A PyPop configuration file is not generated by
BDtoPyPop().</p>
<pre><code>Generating a list object of two PyPop datasets from the HLA_data dataset bundled with BIGDAWG.
HLAdata.PP &lt;- BDtoPyPop(sHLAdata,&quot;BDHLA&quot;,FALSE)

Generating a pair PyPop data files from the HLA_data dataset bundled with BIGDAWG.
BDtoPyPop(sHLAdata,&quot;BDHLA&quot;,TRUE)</code></pre>
<hr />
<div align="center">
<p>END OF VIGNETTE</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
